// testcafe/lib/api/request-hooks/request-mock.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const hook_1 = __importDefault(require("./hook"));
const testcafe_hammerhead_1 = require("testcafe-hammerhead");
const runtime_1 = require("../../errors/runtime");
const types_1 = require("../../errors/types");
const warning_message_1 = __importDefault(require("../../notifications/warning-message"));
class RequestMock extends hook_1.default {
    constructor() {
        super([]);
        this._pendingRequestFilterRuleInit = null;
        this._mocks = new Map();
    }
    async onRequest(event) {
        const mock = this._mocks.get(event._requestFilterRule.id);
        await event.setMock(mock);
    }
    async onResponse(event) {
        var _a;
        if (event.isSameOriginPolicyFailed)
            (_a = this._warningLog) === null || _a === void 0 ? void 0 : _a.addWarning(warning_message_1.default.requestMockCORSValidationFailed, RequestMock.name, event._requestFilterRule);
    }
    // API
    onRequestTo(requestFilterRuleInit) {
        if (this._pendingRequestFilterRuleInit)
            throw new runtime_1.APIError('onRequestTo', types_1.RUNTIME_ERRORS.requestHookConfigureAPIError, RequestMock.name, "The 'respond' method was not called after 'onRequestTo'. You must call the 'respond' method to provide the mocked response.");
        this._pendingRequestFilterRuleInit = requestFilterRuleInit;
        return this;
    }
    respond(body, statusCode, headers) {
        if (!this._pendingRequestFilterRuleInit)
            throw new runtime_1.APIError('respond', types_1.RUNTIME_ERRORS.requestHookConfigureAPIError, RequestMock.name, "The 'onRequestTo' method was not called before 'respond'. You must call the 'onRequestTo' method to provide the URL requests to which are mocked.");
        const mock = new testcafe_hammerhead_1.ResponseMock(body, statusCode, headers);
        const rule = new testcafe_hammerhead_1.RequestFilterRule(this._pendingRequestFilterRuleInit);
        this._requestFilterRules.push(rule);
        this._mocks.set(rule.id, mock);
        this._pendingRequestFilterRuleInit = null;
        return this;
    }
}
function createRequestMock() {
    return new RequestMock();
}
exports.default = createRequestMock;
module.exports = exports.default;
